### ISSUE TITLE
Issues with Decimal128 type

### ISSUE THREAD

**Issue by HiiiiD** [Automattic/mongoose#12431]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Mongoose version

6.6.1

### Node.js version

18.2.0

### MongoDB server version

5.x

### Description

Adding a field to a Mongoose Schema with the type `Schema.Types.Decimal128` breaks all the typings of `model`, including the ones of the other fields of the Mongoose Schema.
Make sure to use a version of Typescript < 4.8.x due to the fact that otherwise the build breaks.

### Steps to Reproduce

For reference this is the `tsconfig.json` that I used to compile
```json
{
    "compilerOptions": {
        "declaration": true,
        "outDir": "./build",
        "module": "CommonJS"
    },
    "exclude": [
        "node_modules",
        "build"
    ]
}
```

```ts
import { model, Schema } from 'mongoose';

const testSchema = new Schema({
    testDate: { type: Date }
});

const TestModel = model('Test', testSchema);
```
As expected `TestModel` has the following type
```ts
declare const TestModel: import("mongoose").Model<{
    testDate?: Date;
}, {}, {}, {}, Schema<any, import("mongoose").Model<any, any, any, any, any>, {}, {}, {}, {}, "type", {
    testDate?: Date;
}>>;
```

If we change our schema just a little bit, introducing `Schema.Types.Decimal128` it breaks every type.

```ts
import { model, Schema } from 'mongoose';

const testSchema = new Schema({
    testDate: { type: Date },
    testDecimal: { type: Schema.Types.Decimal128 },
});

const TestModel = model('Test', testSchema);
```

This results in
```ts
declare const TestModel: import("mongoose").Model<{
    testDate?: {
        [Symbol.toPrimitive]?: {};
        toString: {};
        valueOf: {};
        toLocaleString: {};
        toJSON?: {};
        toDateString?: {};
        toTimeString?: {};
        toLocaleDateString?: {};
        toLocaleTimeString?: {};
        getTime?: {};
        getFullYear?: {};
        getUTCFullYear?: {};
        getMonth?: {};
        getUTCMonth?: {};
        getDate?: {};
        getUTCDate?: {};
        getDay?: {};
        getUTCDay?: {};
        getHours?: {};
        getUTCHours?: {};
        getMinutes?: {};
        getUTCMinutes?: {};
        getSeconds?: {};
        getUTCSeconds?: {};
        getMilliseconds?: {};
        getUTCMilliseconds?: {};
        getTimezoneOffset?: {};
        setTime?: {};
        setMilliseconds?: {};
        setUTCMilliseconds?: {};
        setSeconds?: {};
        setUTCSeconds?: {};
        setMinutes?: {};
        setUTCMinutes?: {};
        setHours?: {};
        setUTCHours?: {};
        setDate?: {};
        setUTCDate?: {};
        setMonth?: {};
        setUTCMonth?: {};
        setFullYear?: {};
        setUTCFullYear?: {};
        toUTCString?: {};
        toISOString?: {};
        getVarDate?: {};
    };
    testDecimal?: {
        toString: {};
        toJSON?: {};
        bytes?: {
            [x: number]: unknown;
            [Symbol.toStringTag]?: unknown;
            [Symbol.iterator]?: {};
            length?: unknown;
            toString: {};
            indexOf?: {};
            lastIndexOf?: {};
            slice?: {};
            valueOf: {};
            includes?: {};
            at?: {};
            toLocaleString: {};
            join?: {};
            every?: {};
            some?: {};
            forEach?: {};
            map?: {};
            filter?: {};
            reduce?: {};
            reduceRight?: {};
            find?: {};
            findIndex?: {};
            entries?: {};
            keys?: {};
            values?: {};
            reverse?: {};
            sort?: {};
            fill?: {};
            copyWithin?: {};
            BYTES_PER_ELEMENT?: unknown;
            buffer?: {
                [Symbol.toStringTag]?: unknown;
                slice?: {};
                byteLength?: unknown;
            } | {
                [Symbol.toStringTag]?: unknown;
                slice?: {};
                [Symbol.species]?: any;
                byteLength?: unknown;
            };
            byteLength?: unknown;
            byteOffset?: unknown;
            set?: {};
            subarray?: {};
            copy?: {};
            write?: {};
            toJSON?: {};
            equals?: {};
            compare?: {};
            writeBigInt64BE?: {};
            writeBigInt64LE?: {};
            writeBigUInt64BE?: {};
            writeBigUint64BE?: {};
            writeBigUInt64LE?: {};
            writeBigUint64LE?: {};
            writeUIntLE?: {};
            writeUintLE?: {};
            writeUIntBE?: {};
            writeUintBE?: {};
            writeIntLE?: {};
            writeIntBE?: {};
            readBigUInt64BE?: {};
            readBigUint64BE?: {};
            readBigUInt64LE?: {};
            readBigUint64LE?: {};
            readBigInt64BE?: {};
            readBigInt64LE?: {};
            readUIntLE?: {};
            readUintLE?: {};
            readUIntBE?: {};
            readUintBE?: {};
            readIntLE?: {};
            readIntBE?: {};
            readUInt8?: {};
            readUint8?: {};
            readUInt16LE?: {};
            readUint16LE?: {};
            readUInt16BE?: {};
            readUint16BE?: {};
            readUInt32LE?: {};
            readUint32LE?: {};
            readUInt32BE?: {};
            readUint32BE?: {};
            readInt8?: {};
            readInt16LE?: {};
            readInt16BE?: {};
            readInt32LE?: {};
            readInt32BE?: {};
            readFloatLE?: {};
            readFloatBE?: {};
            readDoubleLE?: {};
            readDoubleBE?: {};
            swap16?: {};
            swap32?: {};
            swap64?: {};
            writeUInt8?: {};
            writeUint8?: {};
            writeUInt16LE?: {};
            writeUint16LE?: {};
            writeUInt16BE?: {};
            writeUint16BE?: {};
            writeUInt32LE?: {};
            writeUint32LE?: {};
            writeUInt32BE?: {};
            writeUint32BE?: {};
            writeInt8?: {};
            writeInt16LE?: {};
            writeInt16BE?: {};
            writeInt32LE?: {};
            writeInt32BE?: {};
            writeFloatLE?: {};
            writeFloatBE?: {};
            writeDoubleLE?: {};
            writeDoubleBE?: {};
        };
        _bsontype?: import("mongoose").Types.Decimal128;
        inspect?: {};
    };
}, {}, {}, {}, Schema<any, import("mongoose").Model<any, any, any, any, any>, {}, {}, {}, {}, "type", {
    testDate?: Date;
    testDecimal?: import("mongoose").Types.Decimal128;
}>>;
```

It seems that adding the Decimal128 unpacks every type in the Schema.

### Expected Behavior

It should simply be something like:
```ts
declare const TestModel: import("mongoose").Model<{
    testDate?: Date;
    testDecimal?: import("mongoose").Types.Decimal128;
}, {}, {}, {}, Schema<any, import("mongoose").Model<any, any, any, any, any>, {}, {}, {}, {}, "type", {
    testDate?: Date;
    testDecimal?: import("mongoose").Types.Decimal128;
}>>;
```

**Comment by IslandRhythms** [comment]
What are you doing to get that result?

**Comment by hasezoey** [comment]
maybe related to #12352?

**Comment by HiiiiD** [comment]
I think so

**Comment by HiiiiD** [comment]
I've also tried digging down TS and it seems that `infer` breaks typings; I know it's weird but if, for example, we hover over, with VSCode, the types that the Schema creates they're completely fine.

**Comment by hasezoey** [comment]
> I know it's weird but if, for example, we hover over, with VSCode, the types that the Schema creates are completely fine.

for context, what version of tsc is used to build and what version of typescript is used in vscode? (`tsc --version` and `CTRL+P(default) -> Select Typescript version -> see which one has a "dot" infront`)

<details>
<summary>VSCode typescript version example</summary>

![vscode tsc version](https://user-images.githubusercontent.com/10911626/190358015-5d0d234b-28ee-4be1-b730-8ef41545a919.png)

</details>

**Comment by HiiiiD** [comment]
I use the Workspace Version in VSCode, so it's 4.7.4 for both VSCode and tsc

**Comment by HiiiiD** [comment]
> I've also tried digging down TS and it seems that `infer` breaks types; I know it's weird but if, for example, we hover over, with VSCode, the types that the Schema creates are completely fine.

This is what I meant
![image](https://user-images.githubusercontent.com/61231210/190359745-4d701576-6d60-4793-a5bc-0fc3ee70a488.png)

**Comment by hasezoey** [comment]
i investigated a bit for https://github.com/Automattic/mongoose/pull/12450#issuecomment-1255537212 and found that out of whatever reason, typescript does not like having a `T extends Type<infer T1, infer T2> ? { t1: T1, t2: T2 }[alias] : never`, it somehow changes the type, but doing `T extends Type<any, infer T2> ? T2 : never`, it gives the correct type

or in more mongoose-types terms, the following will somehow change the type:
```ts
type ObtainSchemaGeneric<TSchema, alias extends 'EnforcedDocType' | 'M' | 'TInstanceMethods' | 'TQueryHelpers' | 'TVirtuals' | 'TStaticMethods' | 'TPathTypeKey' | 'DocType'> =
  TSchema extends Schema<infer EnforcedDocType, infer M, infer TInstanceMethods, infer TQueryHelpers, infer TVirtuals, infer TStaticMethods, infer TPathTypeKey, infer DocType>
    ? {
      EnforcedDocType: EnforcedDocType;
      M: M;
      TInstanceMethods: TInstanceMethods;
      TQueryHelpers: TQueryHelpers;
      TVirtuals: TVirtuals;
      TStaticMethods: TStaticMethods;
      TPathTypeKey: TPathTypeKey;
      DocType: DocType;
    }[alias]
    : unknown;

type InferSchemaType<SchemaType> = ObtainSchemaGeneric<SchemaType, 'DocType'> ;

const Schema2 = new Schema({ // correct 8th generic (DocType)
  createdAt: { type: Date, required: true },
  decimalValue: { type: Schema.Types.Decimal128, required: true }
});

const t = {} as InferSchemaType<typeof Schema2>; // somehow has a modified DocType
```

while the following returns the type intact:
```ts
type ObtainDocType<TSchema> =
  TSchema extends Schema<any, any, any, any, any, any, any, infer DocType>
    ? DocType
    : unknown;

type InferSchemaType<SchemaType> = ObtainDocType<SchemaType>;

const Schema2 = new Schema({ // correct 8th generic (DocType)
  createdAt: { type: Date, required: true },
  decimalValue: { type: Schema.Types.Decimal128, required: true }
});

const t = {} as InferSchemaType<typeof Schema2>; // DocType is returned without modification
```

@Uzlopak @vkarpov15 @IslandRhythms  what do you think about that?

**Comment by HiiiiD** [comment]
I've tried to add/remove conditions to `ResolvePathType` in order to see where it starts to break, and apparently removing `PathValueType extends Record<string, any> ? ObtainDocumentType<PathValueType, any, TypeKey>` fixes my issue, but it has clearly broken tests that depend on it

I don't understand why without this condition everything works, and adding it breaks my use case

**Comment by vkarpov15** [comment]
I took a look and confirmed that this issue was fixed in 6.7.0 with #12352. I added a test case that covers the issue described in OP and it is passing.
